src/cluster_linearize.h

```diff
@@ -738,6 +804,233 @@ std::pair<std::vector<ClusterIndex>, bool> Linearize(const DepGraph<SetType>& de
+        /** The combined fee/size of transactions in the group. Fee is negated in even passes. */
+        FeeFrac feerate;
+    };
+
+    // As an example, consider the state corresponding to the linearization [1,0,3,2], with
+    // groups [1,0,3] and [2], in an odd pass. The linked lists would be:
+    //
+    //                                        +-----+
+    //                                 0<-P-- | 0 S | ---\     Legend:
+    //                                        +-----+    |
+    //                                           ^       |     - digit in box: entries index
+    //             /--------------F---------+    G       |       (note: one more than tx value)
+    //             v                         \   |       |     - S: sentinel group
+    //          +-----+        +-----+        +-----+    |          (empty feerate)
+    //   0<-P-- | 2   | <--P-- | 1   | <--P-- | 4 T |    |     - T: tail transaction, contains
+    //          +-----+        +-----+        +-----+    |          fields beyond prev_tv.
+    //                                           ^       |     - P: prev_tx reference
+    //                                           G       G     - F: first_tx reference
+    //                                           |       |     - G: prev_group reference
+    //                                        +-----+    |
+    //                                 0<-P-- | 3 T | <--/
+    //                                        +-----+
+    //                                         ^   |
+    //                                         \-F-/
+    //
+    // During an even pass, the diagram above would correspond to linearization [2,3,0,1], with
+    // groups [2] and [3,0,1].
+
+    std::vector<TxEntry> entries(linearization.size() + 1);
+
+    // Perform two passes over the linearization.
+    for (int pass = 0; pass < 2; ++pass) {
+        int rev = !(pass & 1);
+        // Construct a sentinel group, identifying the start of the list.
+        entries[0].prev_group = 0;
+        Assume(entries[0].feerate.IsEmpty());
+
+        // Iterate over all elements in the existing linearization.
+        for (ClusterIndex i = 0; i < linearization.size(); ++i) {
+            // Even passes are from back to front; odd passes from front to back.
+            ClusterIndex idx = linearization[rev ? linearization.size() - 1 - i : i];
+            // Construct a new group containing just idx. In even passes, the meaning of
+            // parent/child and high/low feerate are swapped.
+            ClusterIndex cur_group = idx + 1;
+            entries[cur_group].group = SetType::Singleton(idx);
+            entries[cur_group].deps = rev ? depgraph.Descendants(idx): depgraph.Ancestors(idx);
+            entries[cur_group].feerate = depgraph.FeeRate(idx);
+            if (rev) entries[cur_group].feerate.fee = -entries[cur_group].feerate.fee;
+            entries[cur_group].prev_tx = 0; // No previous transaction in group.
```

Comment:
except for here, I think all uses of 0 are sentinel references, could add a constant to denote that meaning?